metricbeat/module/system/diskio/diskstat_linux.go

```diff
@@ -34,7 +34,81 @@ func Get_CLK_TCK() uint32 {
 
 // IOCounters should map functionality to disk package for linux os.
 func IOCounters(names ...string) (map[string]disk.IOCountersStat, error) {
-	return disk.IOCounters(names...)
+	stats, err := disk.IOCounters(names...)
+	if err != nil {
+		return nil, err
+	}
+	//Process `stats`, as `names` might be empty
+	topCounters := separateTopLevelCounters(stats)
+	stats["summary"] = summaryIOCounter(topCounters)
+	return stats, nil
+}
+
+func summaryIOCounter(stats map[string]disk.IOCountersStat) disk.IOCountersStat {
+	sum := disk.IOCountersStat{}
+	sum.Name = "summary"
+
+	for _, counter := range stats {
+		sum.ReadCount += counter.ReadCount
+		sum.ReadBytes += counter.ReadBytes
+		sum.ReadTime += counter.ReadTime
+		sum.WriteCount += counter.WriteCount
+		sum.WriteBytes += counter.WriteBytes
+		sum.WriteTime += counter.WriteTime
+		sum.IoTime += counter.IoTime
+		sum.MergedReadCount += counter.MergedReadCount
+		sum.MergedWriteCount += counter.MergedWriteCount
+		sum.WeightedIO += counter.WeightedIO
+	}
+
+	return sum
+}
+
+// separateTopLevelCounters will return a map of top level counters to ,
+// and it removes only aggregate counters from the passed map `stats`.
+// for example, it adds `hda` to a new map, and will not remove `hda`
+// if there is no sub counters such as `hda1`
+func separateTopLevelCounters(stats map[string]disk.IOCountersStat) map[string]disk.IOCountersStat {
+	separated := map[string]disk.IOCountersStat{}
+	for name := range stats {
```

Comment:
This should work with keys in any order, I guess. For example, if `sda1` is before `sda`, it will enter `if` in L98, and `sda` would be moved from `stats` to `separated` then. : )

Reply:
oh, you're right, I wrote this before the longer comment but if you make changes like the suggestions there then this one is no longer relevant :-)